lib/mongo/server/connection_pool.rb

```diff
@@ -326,20 +326,24 @@ def check_out
       # @since 2.9.0
       def check_in(connection)
         @lock.synchronize do
-          unless @checked_out_connections.include?(connection)
-            raise ArgumentError, "Trying to check in a connection which is not currently checked out by this pool: #{connection}"
+          unless connection.pool == self
+            raise ArgumentError, "Trying to check in a connection which was not checked out by this pool: #{connection} checked out from pool #{connection.pool} (for #{self})"
           end
 
-          @checked_out_connections.delete(connection)
-
           # Note: if an event handler raises, resource will not be signaled.
           # This means threads waiting for a connection to free up when
           # the pool is at max size may time out.
           # Threads that begin waiting after this method completes (with
           # the exception) should be fine.
-          publish_cmap_event(
-            Monitoring::Event::Cmap::ConnectionCheckedIn.new(@server.address, connection.id)
-          )
+
+          if @checked_out_connections.include?(connection)
+            # It is possible the connection was already checked in; we only
```

Comment:
What is the use case for checking in a connection twice? I am inclined to think this should be an error.

Reply:
This is not a strong case for it, but it would just be more tolerant of connections that are checked out without `with_connection`, which could accidentally be checked in twice if the connection reference is not handled carefully. It makes more sense to error, so I will change it.